libsks/src/pkcs11_token.c

```diff
@@ -208,6 +209,10 @@ CK_RV sks_ck_token_mechanism_ids(CK_SLOT_ID slot,
 			  &ctrl, sizeof(ctrl), NULL, 0, outbuf, &outsize);
 	if (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL)
 		*count = outsize / sizeof(uint32_t);
+	if(!mechanisms && rv == CKR_BUFFER_TOO_SMALL) {
+		rv = CKR_OK;
+		goto bail;
+	}
```

Comment:
I think the code at line 204 should also be changed. `*count` maybe meaningless if caller queries the number of available mechanisms.
Should be something like:
```c
	rv = ck_invoke_ta(NULL, SKS_CMD_CK_MECHANISM_IDS,
			  &ctrl, sizeof(ctrl), NULL, 0, NULL, &outsize);

	if (rv != CKR_BUFFER_TOO_SMALL)
		return CKR_FUNCTION_FAILED;

	if (!mechanisms) {
		*count = outsize / sizeof(uint32_t);
		return CKR_OK;
	}
	if (*count < outsize / sizeof(uint32_t)) {
		*count = outsize / sizeof(uint32_t);
		return rv;
	}

	/* Get the mechanisms IDs now */
	outbuf = malloc(outsize);
 	if (!outbuf)
 		return CKR_HOST_MEMORY;

	rv = ck_invoke_ta(NULL, SKS_CMD_CK_MECHANISM_IDS,
 			  &ctrl, sizeof(ctrl), NULL, 0, outbuf, &outsize);

	if (rv == CKR_OK) {
		if (sks2ck_mechanism_type_list(mechanisms, outbuf, *count)) {
			LOG_ERROR("unexpected bad mechanism_type list\n");
			rv = CKR_DEVICE_ERROR;
		}
	}

	free(outbuf);
	return rv;
```